### DISCUSSION TITLE
File uploads?

### DISCUSSION THREAD

**Discussion by silasb** [jfyne/live#47] [General]
Was playing around with this project and wanted to be able to do file uploads.  My form submits the event, but I'm noticing that the event is missing all the data. 

```
{"t":"upload","i":1,"d":{"myFile":{}}}
```

I just skimmed the project, but from what I can tell, you'd need to slice the file into smaller chunks to send over WebSocket stream, then reassemble the file on the server side.  It'd probably require reworking `LiveEvent` and `Socket.send` to support this lower level operation?

Maybe I'm missing something and this already works, if so it'd be a good example to have.

**Comment by jfyne** [comment]
Hi yes,

This has been on my list for a while, I haven't done it yet. Sorry.

Now that someone has asked for it I will give it a look. It is definitely a special case, looking at the way Phoenix Liveview does it, they upload the files first and then call the form save method. They have a special bit of state on the socket to monitor uploads. That kind of logic should work for this project.

I will endeavour to get around to it sometime this week.

(Sorry I missed this discussion, I get so many notifications on I sometimes miss things)

**Reply by silasb** [reply]
Thanks, now the feeling about notifications, there's only so much time in the day.  I also looked at how Liveview does it as inspiration.  I tried to implement it and have some successful uploads.  I don't think my solution is the best solution, but I'll throw it up for review when I get it a little farther.  Right now the limitation is delaying calling the `save` method until the upload is complete. 

Couple interesting challenges I had were:
* Sending binary over WS
* Consistent encoding/decoding (I think I used base64)
* Callbacks to handle progress updates when uploading
* Callback to handle when the upload is done
* Consuming uploads within the `save` method

Next steps for me are:
* Calling `save` after `upload` is finished
* Converting my rough JS to TS
* Clean up what I have

As I kept going through this I keep thinking about security concerns... One thing that I keep thinking about is someone being able to time an attack to upload a file to server state prior to you actually uploading. From what I can tell, Phoenix Liveview will use a unique pre-generated ID to essentially upload into.  My solution doesn't do that, but might be required for secure uploads.

I hope to have some time today/tomorrow to hopefully make some progress.

**Reply by jfyne** [reply]
I created an issue here to track this better https://github.com/jfyne/live/issues/48

**Comment by jfyne** [comment]
There is a released version of file uploads in `v0.15.0`.
